numpy/core/src/multiarray/compiled_base.c

```diff
@@ -500,6 +506,37 @@ binary_search(double key, double arr [], npy_intp len)
     if (key > arr[len - 1]) {
         return len;
     }
+    else if (key < arr[0]) {
+        return -1;
+    }
+
+    if (guess < 0)  {
+        guess = 0;
+    }
+    else if (guess >= len - 1) {
+        guess = len - 2;
```

Comment:
Sorry that I hadn't consider very short arrays. I would just put a `if (len
> 16){  }` test around the whole set of tests involving the value of guess,
and not bother trying to optimize for very small arrays.